dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs

```diff
@@ -145,83 +241,228 @@ public void VisitBetween(IMatcherFieldSpec<Document> fieldSpec, string startValu
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) > 0);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) > 0);
+            }
         }
 
         public void VisitGreaterThanOrEquals(IMatcherFieldSpec<Document> fieldSpec, long value)
         {
-            _isMatch = GetLongStream(fieldSpec).Any(v => v >= value);
+            var values = GetLongStream(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = values.Any(v => v >= value);
+            }
         }
 
         public void VisitGreaterThanOrEquals(IMatcherFieldSpec<Document> fieldSpec, string value)
         {
-            string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
+            var values = GetStringValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) >= 0);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) >= 0);
+            }
         }
 
         public void VisitExists(IMatcherFieldSpec<Document> fieldSpec)
         {
-            _isMatch = GetFieldValues(fieldSpec).Any();
+            var values = GetFieldValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = false;
+            }
+            else
+            {
+                _isMatch = values.Any();
+            }
         }
 
         public void VisitEmpty(IMatcherFieldSpec<Document> fieldSpec)
         {
-            _isMatch = !GetStringValues(fieldSpec).Any(v => v.Length > 0);
+            var values = GetStringValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = !values.Any(v => v.Length > 0);
+            }
         }
 
         public void VisitOr(IEnumerable<Filter<IMatcherFieldSpec<Document>>> filters)
         {
+            /*
+             true or true === true
+             true or false === true
+             false or true === true
+             false or false === false
+
+             true or unknown === true
+             false or unknown === unknown
+             unknown or true === true
+             unknown or false === unknown
+
+             unknown or unknown === unknown
+            */
+            var matches = new List<bool?>();
             foreach (Filter<IMatcherFieldSpec<Document>> filter in filters)
             {
                 filter.Invoke(this);
-                if (_isMatch)
-                {
-                    break;
-                }
+                matches.Add(_isMatch);
```

Comment:
We're losing short-circuiting here.

Reply:
public void VisitOr(IEnumerable<Filter<IMatcherFieldSpec<Document>>> filters)
    {
        using (var filterEnumerator = filters.GetEnumerator())
        {
            while (filterEnumerator.MoveNext())
            {
                filterEnumerator.Current.Invoke(this);
                if (_isMatch == true)
                {
                    return;
                }

                if (_isMatch == null)
                {
                    while (filterEnumerator.MoveNext())
                    {
                        filterEnumerator.Current.Invoke(this);
                        if (_isMatch == true)
                        {
                            return;
                        }
                    }

                    _isMatch = null;
                    return;
                }
            }
        }
    }